confidence/formats.py

```diff
@@ -0,0 +1,78 @@
+import json
+import typing
+from abc import abstractmethod
+from dataclasses import dataclass, replace
+from os import PathLike
+from pathlib import Path
+
+import yaml
+
+from confidence.models import unwrap
+
+
+@dataclass(frozen=True)
+class Format(typing.Protocol):
+    suffix: str = ''
+    encoding: str = 'utf-8'
+
+    def load(self, fp: typing.TextIO) -> typing.Any:
+        return self.loads(fp.read())
+
+    @abstractmethod
+    def loads(self, string: str) -> typing.Any:
+        raise NotImplementedError
+
+    def loadf(self, fpath: typing.Union[str, PathLike], encoding: typing.Optional[str] = None) -> typing.Any:
+        with Path(fpath).open('rt', encoding=encoding or self.encoding) as fp:
+            return self.load(fp)
+
+    def dump(self, value: typing.Any, fp: typing.TextIO) -> None:
+        fp.write(self.dumps(value))
+
+    @abstractmethod
+    def dumps(self, value: typing.Any) -> str:
+        raise NotImplementedError
+
+    def dumpf(
+        self, value: typing.Any, fname: typing.Union[str, PathLike], encoding: typing.Optional[str] = None
+    ) -> None:
+        with Path(fname).open('wt', encoding=encoding or self.encoding) as fp:
+            return self.dump(value, fp)
+
+    def __call__(self, **kwargs: typing.Any) -> 'Format':  # TODO: replace with typing.Self for Python 3.11+
+        return replace(self, **kwargs)
+
+
+@dataclass(frozen=True)
+class _JSONFormat(Format):
+    suffix: str = '.json'
+
+    def loads(self, string: str) -> typing.Any:
+        return json.loads(string)
+
+    def dumps(self, value: typing.Any) -> str:
+        return json.dumps(unwrap(value))
+
+
+@dataclass(frozen=True)
+class _YAMLFormat(Format):
+    suffix: str = '.yaml'
```

Comment:
If I'm correct its either `.yaml` or `yml`, was that always the case?

Reply:
The 'current' implementation uses "extension" without the dot. As this PR kind of moves things around anyways, I figured I'd take a look at the standard library, where this concept is commonly referred to as a "suffix", *including* the dot. So anyone wanting to use `yml` file extensions could override the default and use `YAML(suffix='.yml')` as the format to be loaded (just `yml` would try to read files like `/etc/exampleyml`).